File: InvoiceCalculatorPython/app/services/InvoiceService.py
from app.repositories.ProductRepository import ProductRepository
from app.schemas.InvoiceSchemas import InvoiceRequest, InvoiceResponse, InvoiceItemResponse
from fastapi import HTTPException


class InvoiceService:
    def __init__(self, product_repository: ProductRepository):
        self.product_repository = product_repository

    def calculate_invoice(self, request: InvoiceRequest) -> InvoiceResponse:
        item_responses = []
        total = 0.0

        for item in request.items:
            product = self.product_repository.find_by_id(item.productId)

            if not product:
                raise HTTPException(
                    status_code=404,
                    detail=f"Product not found with id: {item.productId}"
                )

            if product.stock < item.quantity:
                raise HTTPException(
                    status_code=400,
                    detail=f"Insufficient stock for product: {product.name}"
                )

            subtotal = product.price * item.quantity
            total += subtotal

            item_response = InvoiceItemResponse(
                productId=product.id,
                productName=product.name,
                quantity=item.quantity,
                unitPrice=product.price,
                subtotal=subtotal
            )

            item_responses.append(item_response)

        return InvoiceResponse(items=item_responses, total=total)

File: InvoiceCalculatorJava/src/main/java/com/uniandes/InvoiceCalculatorJava/services/InvoiceService.java
package com.uniandes.InvoiceCalculatorJava.services;

import com.uniandes.InvoiceCalculatorJava.dto.InvoiceItemRequest;
import com.uniandes.InvoiceCalculatorJava.dto.InvoiceItemResponse;
import com.uniandes.InvoiceCalculatorJava.dto.InvoiceRequest;
import com.uniandes.InvoiceCalculatorJava.dto.InvoiceResponse;
import com.uniandes.InvoiceCalculatorJava.models.Product;
import com.uniandes.InvoiceCalculatorJava.repositories.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class InvoiceService {

    @Autowired
    private ProductRepository productRepository;

    public InvoiceResponse calculateInvoice(InvoiceRequest request) {
        List<InvoiceItemResponse> itemResponses = new ArrayList<>();
        double total = 0.0;

        for (InvoiceItemRequest item : request.getItems()) {
            Product product = productRepository.findById(item.getProductId())
                    .orElseThrow(() -> new RuntimeException("Product not found with id: " + item.getProductId()));

            if (product.getStock() < item.getQuantity()) {
                throw new RuntimeException("Insufficient stock for product: " + product.getName());
            }

            double subtotal = product.getPrice() * item.getQuantity();
            total += subtotal;

            InvoiceItemResponse itemResponse = new InvoiceItemResponse(
                    product.getId(),
                    product.getName(),
                    item.getQuantity(),
                    product.getPrice(),
                    subtotal
            );

            itemResponses.add(itemResponse);
        }

        return new InvoiceResponse(itemResponses, total);
    }
}
